### ISSUE TITLE
NumericalityValidator equal_to false positives with BigDecimal attributes

### ISSUE THREAD

**Issue by gabetax** [rails/rails#12134]
Given this model:

```ruby
class Payment
  include ActiveModel::Model

  attr_accessor :amount

  validates :amount, numericality: {
    equal_to: ->(record) { record.amount }
  }
end
```

we should expect the numericality validation to always pass with a value for `amount`. However, it can fail if the amount column is a BigDecimal (which often happens as ActiveRecord represents `decimal` database columns with the BigDecimal type):

```
[7] pry(main)> payment = Payment.new amount: BigDecimal.new('706.06')
=> #<Payment:0x00000004c95c08 @amount=#<BigDecimal:4c95c58,'0.70606E3',18(18)>>
[8] pry(main)> payment.valid?
=> false
[9] pry(main)> payment.errors.to_a
=> ["Amount must be equal to 706.06"]
[10] pry(main)> payment.amount
=> #<BigDecimal:4c95c58,'0.70606E3',18(18)>
[11] pry(main)> payment.amount.to_f
=> 706.06
```

This appears to happen because NumericalityValidator takes the raw string value of :amount, and attempts to re-parse it to a Float: https://github.com/rails/rails/blob/4-0-stable/activemodel/lib/active_model/validations/numericality.rb#L27

And as it turns out, ruby doesn't compare Float and BigDecimal reliably:

```ruby
(0..9).each do |i|
  s = "706.0#{i}"
  puts s
  puts s.to_f == BigDecimal.new(s)
end
```

```
706.00
true
706.01
true
706.02
true
706.03
true
706.04
true
706.05
true
706.06
false
706.07
false
706.08
true
706.09
true
```
(output the same on 1.9.3 and 2.0.0-p247)

I haven't started trudging around the I'm not sure if this should be accepted under the general "Floats are imprecise", or if this is an actual ruby-core bug, but there are a few considerations:

1. Is it sensible to [always re-parse the values to a float](https://github.com/rails/rails/blob/4-0-stable/activemodel/lib/active_model/validations/numericality.rb#L64), and arguably flakey type for comparisons?
2. If so, should we call `.to_f` on the `option_value` as well, so we're comparing the same types?

**Comment by morgoth** [comment]
@gabetax I cannot reproduce it on current master (4.2).

**Comment by chazu** [comment]
+1 on this

**Comment by sgrif** [comment]
Fixed by #19851
